ledger/src/blockstore_meta.rs

```diff
@@ -269,6 +304,339 @@ impl ShredIndex {
+    {
+        use serde::ser::SerializeTuple;
+
+        let mut tuple = serializer.serialize_tuple(2)?;
+        // Purpose:
+        // - Serialize `self.index` (a `[u64; NUM_U64_PER_SHRED_INDEX_V2]` array) into a fixed-size buffer as little-endian bytes.
+        //
+        // SAFETY: This is safe because:
+        // 1. Memory Layout:
+        // - The buffer is sized to `std::mem::size_of::<[u64; NUM_U64_PER_SHRED_INDEX_V2]>()`, matching the source data size.
+        // - Each `u64` is serialized into 8 bytes, ensuring no padding or overflow.
+        //
+        // 2. Pointer Safety:
+        // - `dst` is derived from `buffer.as_mut_ptr()` and points to valid, uninitialized memory.
+        // - The offsets `dst.add(i * 8)` are always within bounds because:
+        //   - `i` ranges from `0` to `NUM_U64_PER_SHRED_INDEX_V2 - 1`.
+        //   - The total size of the buffer is sufficient to accommodate all writes.
+        //
+        // 3. Alignment:
+        // - `dst` is `u8`-aligned from `buffer.as_mut_ptr()`.
+        // - Writing 8 bytes per iteration (`[u8; 8]` from `to_le_bytes()`) respects alignment requirements.
+        //
+        // 4. Initialization:
+        // - Each 8-byte chunk is written exactly once using `std::ptr::write`, fully initializing the buffer.
+        // - After the loop, the buffer is guaranteed to be completely initialized, making `assume_init()` safe.
+        //
+        // Notes:
+        // - Serialization uses little-endian order for portability.
+        let buffer = unsafe {
+            let mut buffer =
+                MaybeUninit::<[u8; std::mem::size_of::<ShredIndexV2Inner>()]>::uninit();
+            let dst = buffer.as_mut_ptr() as *mut u8;
+            for (i, &word) in self.index.iter().enumerate() {
+                // Explicitly convert to LE bytes
+                std::ptr::write(dst.add(i * 8) as *mut [u8; 8], word.to_le_bytes());
+            }
+            buffer.assume_init()
+        };
+        tuple.serialize_element(serde_bytes::Bytes::new(&buffer))?;
+        tuple.serialize_element(&self.num_shreds)?;
+        tuple.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for ShredIndexV2 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let (bytes, num_shreds) = <(&[u8], usize)>::deserialize(deserializer)?;
+        let total_size = std::mem::size_of::<ShredIndexV2Inner>();
+        // Accept input smaller than our current fixed array size to maintain compatibility with older
+        // versions that had fewer shreds per slot. This is safe because smaller sets are stored in
+        // our fixed array with any remaining space automatically zeroed. This approach ensures we can
+        // read data from any previous version while enforcing a maximum size limit.
+        if bytes.len() > total_size {
+            return Err(serde::de::Error::custom("input too large"));
+        }
```

Comment:
What if we have to reduce `MAX_DATA_SHREDS_PER_SLOT`?
Like we did with `MAX_CODE_SHREDS_PER_SLOT`: https://github.com/solana-labs/solana/pull/30543

Reply:
As mentioned in slack, if this is a real concern, we can add a `_padding: u8` or whatever field to `ShredIndexV2`, then `BTreeSet<u64>` can never deserialize to it because it'll either run out of data at the end missing 6 bytes, or have 1 byte trailing data after deserializing

Reply:
Given that we currently _deserialize_ so often, avoiding a heap allocation would be preferable so I'd go for solution 1. It's true that it's more wasteful but it wastes a bit of stack memory, which is effectively free to allocate as long as we don't accidentally start copying the index around.

Reply:
> What if we have to reduce MAX_DATA_SHREDS_PER_SLOT?

FWIW - any change (increase or decrease) will have to be a SIMD this time unlike the PR you linked. So, while it will be nice to have a general solution, toggling the constant will be a much more involved task from here on out.

Given that it would have to be a SIMD (and thus feature gated), I think we would already need to bubble in the ability to determine whether the feature is enabled (which will be a mess to do 😬). If we already wire up the ability to determine if the feature is activated (or really, the activation slot so we can apply the correct limit for any given slot), then it doesn't seem to be much worse to determine if we should deserialize into ` ShredIndexV2` or a `ShredIndexV3`

Reply:
> > What if we have to reduce MAX_DATA_SHREDS_PER_SLOT?
> 
> FWIW - any change (increase or decrease) will have to be a SIMD this time unlike the PR you linked.

Yeah normally we need a feature-gate and simd.
But https://github.com/solana-labs/solana/pull/30543 was done in response to an outage.
If that outage happens today we will probably go the same route with no simd or feature-gate.


Anyways, I think the first solution to decouple from `MAX_DATA_SHREDS_PER_SLOT` is good enough.